### PR TITLE
implement Event directly with wait_task_rescheduled rather than ParkingLot

### PR DESCRIPTION
**Pull request by belm0**
raised in #1944

### CODE REVIEW

File: trio/_sync.py [hunk 1]
```diff
@@ -59,7 +65,15 @@ async def wait(self):
         if self._flag:
             await trio.lowlevel.checkpoint()
         else:
-            await self._lot.park()
+            task = _core.current_task()
+            self._tasks.add(task)
+            task.custom_sleep_data = self
```
Comment (richardsheridan) [inline trio/_sync.py:70]: Why use custom sleep data here? Can't you just access `self._tasks` directly in the closure?

File: trio/_sync.py [hunk 2]
```diff
@@ -47,8 +51,10 @@ def is_set(self):
     @enable_ki_protection
     def set(self):
         """Set the internal flag value to True, and wake any waiting tasks."""
-        self._flag = True
-        self._lot.unpark_all()
+        if not self._flag:
+            self._flag = True
+            for task in self._tasks:
+                _core.reschedule(task)
```
Comment (njsmith) [inline trio/_sync.py:57]:
```suggestion
                _core.reschedule(task)
            self._tasks.clear()
```

File: trio/_sync.py (hunk 1 above)
Comment (belm0) [reply trio/_sync.py:70]: good point

I've copied this code from ParkingLot, so I guess I'll apply the same simplification there

Comment (njsmith) [reply trio/_sync.py:70]: `ParkingLot` needs `custom_sleep_data` b/c of the "requeueing" functionality that condition variables need -- see `ParkingLot.repark`

Comment (belm0) [reply trio/_sync.py:70]: I see, ParkingLot uses `custom_sleep_data` to support `repark()`

**Comment by belm0** [comment]
ready to go (unless it needs a newsfragment)

suggest "squash and merge" option for merge

**Pull request merged**